### ISSUE TITLE
Should SPACE to be allowed for use as an activation event for the `<a>` element?

### ISSUE THREAD

**Issue by MelSumner** [w3c/aria#1701]
Use case:
I've been writing a call-to-action component. It visually presents as a button; when the developer invokes it, they provide either an action or a URL/route. Based on what they provide, the component renders a  button element (if action) or link (if URL/route). 

To a user with assistive technology, this does the correct thing and everything is peachy there. 

For the user who is a keyboard-only user, this pattern falls down because it visually presents as a button, but, unlike the button (or other form controls) it does not respond to `SPACE` (spacebar).

## Research

In reading the [UI Events KeyboardEvent key Values](https://www.w3.org/TR/uievents-key/) specification, I find no mention of the `SPACE` key, specifically except as a side note in the [Whitespace Keys](https://www.w3.org/TR/uievents-key/#keys-whitespace) section: 

> The space or spacebar key is encoded as " ".

I checked [Navigation Keys](https://www.w3.org/TR/uievents-key/#keys-navigation); since when a a link (`<a>` element) has focus, pressing the `SPACE` key will scroll that area to the top of the user's viewport (but not consistently). However, it is not listed as a Navigation Key item. 

Additional finds: 

1. `ENTER` and `TAB` are are listed both in [Control Characters](https://www.w3.org/TR/uievents-key/#control) **and** [Whitespace Keys](https://www.w3.org/TR/uievents-key/#keys-whitespace)
2. I cannot find any mention of what the `SPACE` key should do  in the HTML specification. It seems to more generically refer to [event target activation behavior](https://dom.spec.whatwg.org/#eventtarget-activation-behavior) and [user interaction task source](https://html.spec.whatwg.org/multipage/webappapis.html#user-interaction-task-source). 
3. UAAG outlines [focus mechanism requirements](https://www.w3.org/WAI/UA/2010/ED-UAAG20-20100308/#gl-focus-mechanism), is there a specific interpretation for these?

## How it works now

I need to test more browser/OS combinations; this is current Chrome on current macOS. 

If I am on a page (let's say https://www.w3.org/WAI/WCAG21/quickref/) and I `TAB` down the page, I could have a link focused that is near the bottom of my viewport. Once I start to leave the viewport, the page scrolls automatically. However, if I have not left the viewport yet (and I'm still near the bottom of the viewport), when I press `SPACE`, the browser scrolls the page for me, so that the focused link is at the top of my viewport...sort of. It ends up being actually just out of the top of my visible viewport, (which might be a bug itself) but the content right under that link is definitely at the top of my viewport. 

## Request for consideration

Because users are able to scroll the page in other ways, (ARROW keys, HOME, END, PAGE UP, PAGE DOWN)
Because adding the `SPACE` functionality to the `<a>` element would resolve a keyboard-only user gap for something that needs to be a link but visually presented as a button (via styles), 

Would we be willing to allow `SPACE` to be used for activation of links in the same way that it can activate a button?

**Comment by chlane** [comment]
@MelSumner could this be done as a widget in the authoring guidelines? Then we could add the guidance to support the spacebar in addition to the Enter key there. We would also need a widget name that made sense.

**Comment by MelSumner** [comment]
@chlane in theory, it is one potential outcome. However, I think I am suggesting something even stronger; that by default, the `a` element should respond to `SPACE` the same way the `button` element does (and, let's put that specifically in writing in the appropriate spec). WDYT?

**Comment by chlane** [comment]
@MelSumner I think it would be great if the spacebar activated the <a> element or an ARIA link.

**Comment by scottaohara** [comment]
>For the user who is a keyboard-only user, this pattern falls down.

What does 'falls down' mean here?  It's not like it's utterly unusable?  Do we have data to back up how often people use space vs enter to try to activate elements that look like buttons, and would help us determine the severity of this issue? Per at least the initial topic/proposal, it seems like a big swing to change the behavior due to instances of CTA controls, but a user would otherwise know not to press space for links that were styled closer to their defaults, unless they wanted to quickly scroll the page for reading purposes.

Playing devil's advocate (or continuing to i guess), the reverse - a button styled to look like a link should no longer be actionable via the space key, but instead scroll the browser viewport - should hold similar weight as a potential outcome.  I'm absolutely not advocating for that, but want to call out that they are two sides of the same solution.

My major concern here is taking away a long standing behavior from links because _some_ links, in the grand scope of all the links on the internet, are styled to look differently.  Yes, there are other ways to scroll a page if taking away this feature, but they are slower (arrow keys) or not equivalent or even available on all keyboard types (page up/down).

Those will be my primary questions to address in the larger discussion :)

**Comment by JAWS-test** [comment]
I don't really understand the question because, as far as I know, neither the ARIA specification nor the HTML specification specifies how certain elements can be operated. This seems to me to be left to the respective operating systems and browsers. Of course there are conventions, but no binding rules. The conventions are mentioned in https://www.w3.org/TR/wai-aria-practices-1.1/, for example, but are not defined there. I.e. the only thing we could describe in [ARIA Authoring Practices](https://w3c.github.io/aria-practices/examples/link/link.html) would be that role=link should be optional or necessarily operable with SPACE. If this is desired, the discussion should perhaps be moved to https://github.com/w3c/aria-practices.

If you design a link like a button (which is not a good idea) you can already make the link activatable with SPACE. This would not be a violation of any specification - but also not required according to WCAG 2.1.1, since WCAG does not require a specific keyboard operation, but only any.

At present, SPACE is identical with PAGE DOWN. This is independent of whether the focus is on a link or not.

**Comment by MelSumner** [comment]
@JAWS-test thank you for giving that extra clarification. 💯 

I think it makes the ask even more crisp and narrower in scope, which will be useful.

**Comment by aleventhal** [comment]
This should be filed on WHATWG/HTML, because it's not about AT mapping.
It's about how UA's support <a>.

This would not be a 100% backward compatible change. It could break those some web apps, somewhere, if they reuse spacebar for something else, and expect that to work when a link has focus. I don't know of any, but the web has everything.

**Comment by MelSumner** [comment]
So I have been testing this out on a bunch of different pages, and I think that we need more feedback. 

I am seeing many instances of "if it is not a form element, spacebar scrolls the page". I think any scrolling I was seeing with a link focused was coincidental; I don't think this is how it works).

Def need more testing to see if it does anything when a link is focused (but, probably more efficient to look at browser code at this point).

**Comment by scottaohara** [comment]
yes.  Space key is not doing special on hyperlinks.  It is performing as expected because hyperlinks don't listen for space key press.

**Comment by JAWS-test** [comment]
@aleventhal 

> This should be filed on WHATWG/HTML, because it's not about AT mapping.

Theoretically it belongs more to WHATWG/HTML, but even there it would be out of place, is my guess. Because where in the WHATWG/HTML is it defined how an HTML element is to be operated? Maybe I missed it, but so far I haven't found that.

**Comment by brennanyoung** [comment]
If an author makes a focused link respond to Space as well as Enter. Does it violate anything but expectations? Any harm here? I've also seen people getting confused when radio buttons fail to respond to Enter, because this is a special case, even though it conforms to the ARIA spec.

I've posted on the APG board calling for a mention of these discrepancies in the page about developing a keyboard interface, because they're not obvious.
